pages/en/lb4/Creating-components.md

```diff
@@ -307,3 +307,297 @@ class EmailComponent {
+    clearInterval(interval);
+    return Promise.resolve();
+  }
+}
+```
+In this example, the `ReportProcessingServer` will load new reports every two
+minutes from the database (using the provided repository), process those entries,
+and then update the database.
+
+### Using your server
+To use your server, simply bind it to your application's context.
+
+```ts
+import {ReportProcessingServer} from '../servers';
+
+export class MyApplication extends Application {
+  super();
+  // Will automatically start when app.start is called
+  this.server(ReportProcessingServer); 
+}
+```
+
+### A more advanced example
+Typically, you'll want server instances that listen for traffic on one or more
+ports (this is why they're called "servers", after all). This leads into a key
+concept to leverage for creating your custom servers.
+
+### Controllers and routing
+LoopBack 4 developers are strongly encouraged to use controllers for their
+modules, and this naturally leads to the concept of routing.
+
+No matter what protocol you intend to use for your custom server, you'll need
+to use some algorithm to determine _which_ controller and function to send
+request data to, and that means you need a router.
+
+For example, consider a "toy protocol" similar to the JSON RPC
+specification (but nowhere near as complete or robust).
+
+The toy protocol will require a JSON payload with three properties: `controller`,
+`method`, and `input`.
+
+An example request would look something like this:
+```json
+{
+  "controller": "GreetController",
+  "method": "basicHello",
+  "input": {
+    "name": "world",
+  }
+}
+```
+
+### Defining the controller
+Here is the `GreetController` and the `Person` type:
+
+{% include code-caption.html content="/src/controllers/greet.controller.ts" %}
+
+```ts
+export class GreetController {
+  basicHello(input: Person) {
+    return `Hello, ${input && input.name || 'World'}!`;
+  }
+
+  hobbyHello(input: Person) {
+    return `${this.basicHello(input)} I heard you like ${input && input.hobby ||
+      'underwater basket weaving'}.`;
+  }
+}
+```
+
+{% include code-caption.html content="/src/models/person.model.ts" %}
+```ts
+// Note that this can also be a class!
+export type Person = {
+  name: string;
+  hobby: string;
+};
+```
+
+### Defining our router and server
+Now that there is a controller to test with, the following code creates a `Router` and `Server`.
+This example uses [express](https://expressjs.com/) and the
+`@types/express` package.
+
+{% include code-caption.html content="/src/server/rpc-router.ts" %}
+```ts
+import {RPCServer} from './rpc-server';
+import * as express from 'express';
+import * as parser from 'body-parser';
+
+export class RPCRouter {
+  routing: express.Router;
+  constructor(private server: RPCServer) {
+    this.routing = express.Router();
+    const jsonParser = parser.json();
+    this.routing.post('*', jsonParser, async (request, response) => {
+      console.log(JSON.stringify(request.body));
```

Comment:
What if this line throws an error? The top-level async function will return a rejected promise, which will be ignored because express does not report async middleware/route functions.

Please consider cleaning up this code and making it more robust.

```ts
class {
  constructor(...) {
    this.routing.post('*', jsonParser, (request, response) => {
      this.handleRequest(request, response)
        .catch(err => this.handleError(request, response, error));
    });
  }

  async handleRequest(request, response) {
    console.log(JSON.stringify(request.body));
    // etc.
  }

  handleError(request, response, err) {
     console.error('Cannot handle %s %s:', request.method, request.path, err);
     if (response.headersSent) return;
     response.statusCode = 500;
     response.end();
  });
}
```